File: Api/Api/Services/Predictors/Implementations/PlayerPredictor.cs
using System;
using System.Text.Json;
using Api.Dtos;
using Api.Exceptions;
using Api.Predictors;
using Microsoft.Extensions.Logging;

namespace Api.Services.Predictors.Implementations;

public class PlayerPredictor : Predictor
{
    private readonly ILogger<Predictor> _logger;
    public PlayerPredictor(ILogger<Predictor> logger)
    {
        _logger = logger;
    }
    
    public override int Predict(JsonElement input)
    {
        PlayerInput playerInput;
        try
        { 
            playerInput = input.Deserialize<PlayerInput>();
        }
        catch (Exception e)
        {
            throw new InvalidPredictionInputException();
        }
        
        _logger.LogDebug("Player prediction: {} x {}", playerInput.FirstPlayerId, playerInput.SecondPlayerId);
        return playerInput.FirstPlayerId;
    }
}

File: Api/internal/prediction/predictor/implementations/player_predictor.go
package implementations

import "Api/internal/prediction/predictor"

type PlayerPredictor struct {
	predictor.Predictor
}

func NewPlayerPredictor() predictor.IPredictor {
	return &PlayerPredictor{
		Predictor: predictor.Predictor{},
	}
}

func (cp *PlayerPredictor) Predict() string {
	return "Player!"
}
